closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -1855,3 +1952,526 @@ int OSSL102_RSA_set0_crt_params(RSA *r2, BIGNUM *dmp1, BIGNUM *dmq1, BIGNUM *iqm
+
+    (*env)->ReleasePrimitiveArrayCritical(env, x, nativeX, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, y, nativeY, JNI_ABORT);
+    (*OSSL_BN_CTX_free)(ctx);
+
+    if (0 == ret) {
+        (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    ret = (*OSSL_EC_GROUP_set_generator)(group, generator, nBN, hBN);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+
+    if (0 == ret) {
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    ret = (*OSSL_EC_KEY_set_group)(key, group);
+
+    if (0 == ret) {
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    return (jlong)(intptr_t)key;
+}
+
+/* Free EC Public/Private Key
+ *
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    ECDestroyKey
+ * Signature: (J)I
+ */
+JNIEXPORT jint JNICALL
+Java_jdk_crypto_jniprovider_NativeCrypto_ECDestroyKey
+  (JNIEnv *env, jclass obj, jlong key)
+{
+    EC_KEY *nativeKey = (EC_KEY*)(intptr_t) key;
+    if (NULL == nativeKey) {
+        return -1;
+    }
+    /* no need to call EC_GROUP_free/EC_POINT_free as EC_KEY_free calls them internally */
+    (*OSSL_EC_KEY_free)(nativeKey);
+    return 0;
+}
+
+/* ECDH key agreement, derive shared secret key
+ *
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    ECDeriveKey
+ * Signature: (JJ[BII)I
+ */
+JNIEXPORT jint JNICALL
+Java_jdk_crypto_jniprovider_NativeCrypto_ECDeriveKey
+  (JNIEnv *env, jclass obj, jlong publicKey, jlong privateKey, jbyteArray secret, jint secretOffset, jint secretLen)
+{
+    EC_KEY *nativePublicKey = (EC_KEY*)(intptr_t) publicKey;
+    EC_KEY *nativePrivateKey = (EC_KEY*)(intptr_t) privateKey;
+    unsigned char* nativeSecret = NULL;
+    int ret = 0;
+
+    nativeSecret = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, secret, 0));
+    if (NULL == nativeSecret) {
+        return -1;
+    }
+
+    /* Derive the shared secret */
+    ret = (*OSSL_ECDH_compute_key)((nativeSecret + secretOffset), secretLen, (*OSSL_EC_KEY_get0_public_key)(nativePublicKey), nativePrivateKey, NULL);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, secret, nativeSecret, 0);
+
+    if (0 == ret) {
+        return -1;
+    }
+
+    return (jint)secretLen;
+}
+
+/** Wrapper for OSSL_EC_KEY_set_public_key_affine_coordinates
+ */
+int setECPublicCoordinates(EC_KEY *key, BIGNUM *x, BIGNUM *y, int field) {
+    return (*OSSL_EC_KEY_set_public_key_affine_coordinates)(key, x, y);
+}
```

Comment:
Formatting: this and the next function.

Reply:
Sorry, how should they be formatted instead?